### PR TITLE
Improve completion's type inference for parameters of lambdas being passed as arguments

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1056325 & a [Developer Community issue](https://developercommunity.visualstudio.com/content/problem/893896/intellisense-works-incorrect-with-overloaded-metho.html).

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 1]
```diff
@@ -10104,11 +10104,77 @@ void M(SomeCollection<SomeItem> c)
     }
 }";
 
-            await VerifyItemExistsAsync(markup, "Substring");
+            await VerifyItemIsAbsentAsync(markup, "Substring");
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Why did we _want_ "Substring" showing up in this test? Seems incorrect.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 2]
```diff
@@ -141,10 +153,8 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
                             continue;
                         }
 
-                        type = parameters[ordinalInLambda].Type;
+                        builder.Add(parameters[ordinalInLambda].Type);
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: When this was outside the `if (type.IsDelegateType())`, we would mix in regular non-delegate parameter types as though they were a type argument to a delegate type. So `Action<int>` would unwrap to `int`, and `string` wouldn't unwrap, but would still be included in the results.

**Comment by dpoeschl** [comment]
Can you please take a look @dotnet/roslyn-ide @CyrusNajmabadi @genlu

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 3]
```diff
@@ -38,6 +40,27 @@ public override ImmutableArray<ISymbol> GetSymbols()
...
+        {
+            if (lambdaSyntax.IsKind<ParenthesizedLambdaExpressionSyntax>(SyntaxKind.ParenthesizedLambdaExpression, out var parenthesizedLambdaSyntax))
+            {
+                var parameters = parenthesizedLambdaSyntax.ParameterList.Parameters;
+                if (parameters.Count > ordinalInLambda)
+                {
+                    var parameter = parameters[ordinalInLambda];
+                    if (parameter.Type != null)
+                    {
+                        explicitLambdaParameterType = _context.SemanticModel.GetTypeInfo(parameter.Type, CancellationToken.None).Type;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs]: CancellatoinToken.None?

Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs]:
```suggestion
                        explicitLambdaParameterType = _context.SemanticModel.GetTypeInfo(parameter.Type, _cancellationToken).Type;
```

File: src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationServiceRunner.vb [hunk 4]
```diff
@@ -54,6 +56,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Recommendations
             Return ImmutableArray(Of ISymbol).Empty
         End Function
 
+        Public Overrides Function TryGetExplicitTypeOfLambdaParameter(lambdaSyntax As SyntaxNode, ordinalInLambda As Integer, <NotNullWhen(True)> ByRef explicitLambdaParameterType As ITypeSymbol) As Boolean
+            Dim lambdaExpressionSyntax = DirectCast(lambdaSyntax, LambdaExpressionSyntax)
+            Dim parameters = lambdaExpressionSyntax.SubOrFunctionHeader.ParameterList.Parameters
+            If parameters.Count > ordinalInLambda Then
+                Dim parameterSyntax = parameters(ordinalInLambda)
+                If parameterSyntax.AsClause IsNot Nothing Then
+                    explicitLambdaParameterType = _context.SemanticModel.GetTypeInfo(parameterSyntax.AsClause.Type).Type
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationServiceRunner.vb]:
```suggestion
                    explicitLambdaParameterType = _context.SemanticModel.GetTypeInfo(parameterSyntax.AsClause.Type, _cancellationToken).Type
```

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs (hunk 3 above)
Comment (dpoeschl) [reply src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs]: Fixed. Thanks!

**Pull request merged**